### PR TITLE
Merge pull request #914 from marci4/Issue900

### LINKED ISSUE #900: OnClose is not called when client disconnect
We just upgraded to the latest version. And I noticed that if client process is killed, On server side, onClose is not called. Instead after 30 seconds or so (related to default value in setConnectionLostTimeout), onClose is then called.

Because we were using a old version before, we know, in old version, if we kill the client process, server side, onClose is immediately get called.

Now, it does not do that.

I searched your bug report and found that bug #886 also described the similar situation.

Is this intented change? or there is way for server side to immediately know client socket disconnection?

### PR DESCRIPTION
**Pull request by marci4**
## Description

Due to refactoring when an IOException happens, the WebSocket was always null and therefore close was never called on it.

## Related Issue

Fixes #900

## Motivation and Context

Regression

## How Has This Been Tested?

Currently only a manual test

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by marci4** [comment]
Currently looking for some input on how to test it (see #900) for more!

File: src/main/java/org/java_websocket/exceptions/WrappedIOException.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import org.java_websocket.WebSocket;
+
+import java.io.IOException;
+
+/**
+ * Exception to wrap an IOException and include information about the websocket which had the exception
+ * @since 1.4.1
+ */
+public class WrappedIOException extends Throwable {
```
Comment (PhilipRoman) [inline src/main/java/org/java_websocket/exceptions/WrappedIOException.java]: Should probably extend Exception instead of Throwable

Comment (marci4) [reply src/main/java/org/java_websocket/exceptions/WrappedIOException.java]: My bad

**Comment by daz1ng** [comment]
### 啊啊啊啊

**Comment by MichaelPote** [comment]
Please can this be merged?
It's been 6 months now and this is kind of a big failure of the library to not call onClose.

**Comment by marci4** [comment]
@MichaelPote feel free to write a test for this issue or contribute to the project in another way!

**Comment by MichaelPote** [comment]
@marci4 I've looked through the pull request and the test you wrote, but unfortunately I lack the context and understanding to contribute. It seems to me like you've provided a fix for the issue and written a test (https://gist.github.com/marci4/6b47a9475d440a3dcd0d5212095d5ab3) already?

**Comment by marci4** [comment]
@MichaelPote no, the test is not finished and I have no real idea on how to get this issue tested. The gist is just how far I got

**Comment by marci4** [comment]
@PhilipRoman could you do a review in my changes. I found a way to test this issue!

File: src/test/java/org/java_websocket/issues/Issue900Test.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+            @Override
+            public void onStart() {
+                countServerDownLatch.countDown();
+            }
+        };
+        new Thread(server).start();
+        countServerDownLatch.await();
+        client.connectBlocking();
+        WebSocketImpl websocketImpl = (WebSocketImpl)new ArrayList<WebSocket>(server.getConnections()).get(0);
```
Comment (PhilipRoman) [inline src/test/java/org/java_websocket/issues/Issue900Test.java]: Is there a reason for not writing it like this:
```java
WebSocketImpl websocketImpl = (WebSocketImpl) server.getConnections().get(0)
```

File: src/test/java/org/java_websocket/issues/Issue900Test.java [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+        websocketImpl.setChannel(new ExceptionThrowingByteChannel());
+        server.broadcast("test");
+        countCloseCallsDownLatch.await();
+    }
+    class ExceptionThrowingByteChannel implements WrappedByteChannel {
+
+        @Override
+        public boolean isNeedWrite() {
+            return false;
```
Comment (PhilipRoman) [inline src/test/java/org/java_websocket/issues/Issue900Test.java]: Shouldn't this return true? Looking at https://github.com/TooTallNate/Java-WebSocket/blob/ec0fa03176a316b1f79f2bb208493659f4c70e6b/src/main/java/org/java_websocket/SocketChannelIOHelper.java#L88-L90

it seems that the `writeMore()` isn't actually called in this test. Or are you relying on other methods to throw the exception?

File: src/test/java/org/java_websocket/issues/Issue900Test.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+import java.lang.reflect.Method;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.concurrent.CountDownLatch;
+
+public class Issue900Test {
+
+    CountDownLatch countServerDownLatch = new CountDownLatch(1);
```
Comment (PhilipRoman) [inline src/test/java/org/java_websocket/issues/Issue900Test.java]: Why is this named `countServerDownLatch`? Seems like `serverStartedLatch` would be a better name

File: src/test/java/org/java_websocket/issues/Issue900Test.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    CountDownLatch countCloseCallsDownLatch = new CountDownLatch(1);
```
Comment (PhilipRoman) [inline src/test/java/org/java_websocket/issues/Issue900Test.java]: Same here, I think `serverStoppedLatch` would be a better name.

File: src/test/java/org/java_websocket/issues/Issue900Test.java (hunk 2 above)
Comment (marci4) [reply src/test/java/org/java_websocket/issues/Issue900Test.java]: getConnections() returns a collection

File: src/test/java/org/java_websocket/issues/Issue900Test.java (hunk 3 above)
Comment (marci4) [reply src/test/java/org/java_websocket/issues/Issue900Test.java]: yeah, good point, I relied on the normal read/write call. Changing readMore/writeMore to true will however ensure better behavior

File: src/test/java/org/java_websocket/issues/Issue900Test.java (hunk 5 above)
Comment (marci4) [reply src/test/java/org/java_websocket/issues/Issue900Test.java]: Done

**Pull request merged**